changes/2024-TBD_cache-across-hierarchy-keyrings/change.md

```diff
@@ -0,0 +1,613 @@
+branch-key-digest, we can remove the branch-key-id-length
+parameter in the formula and make it cleaner.
+
+<!-- TODO: Remove this line before merging to mainline -->
+
+This has AWS Cryptographers' approval
+[[ticket]](https://issues.amazon.com/issues/P148158466).
+
+#### Option 2: Update the Encryption cache identifier to include branch-key-id instead of the branch-key-digest=SHA512(branch-key-id)
+
+Not hashing the branch keys will mean that the input
+to the final hash function will be of variable length,
+which is hard to reason about.
+
+#### Option 3: Don't change anything.
+
+This would mean that the formula will remain incoherent.
+
+### Issue 3: Which hashing algorithm (internal and external) should be used in the Hierarchy Keyring cache identifier and the CachingCMM?
+
+Note that according to the
+[cache identifier formulae](#cache-identifier-formulae),
+as of [today](#today), the length of cache identifier for
+BranchKeyMaterials in Hierarchy Keyring is 32 bytes for
+both Encrypt and Decrypt and the length of cache identifier
+for EncryptionMaterials and DecryptionMaterials in the
+CachingCMM is 64 bytes.
+Both Hierarchy Keyring and CachingCMM identifiers currently
+use SHA512, but the Hierarchy Keyring cache identifier is
+truncated to 32 bytes.
+
+#### Option 1 (Recommended): Using SHA512 for both
+
+SHA-512 has a greater range (512 bits) than
+a truncated (to 32 bytes) SHA-512 (256 bits)
+or SHA-384 (384 bits). The greater the range,
+the less likely there is a collision.
+More information about why this is true based
+on the birthday bound can be found
+[here](#birthday-bound-for-sha512-and-sha384).
+
+<!-- TODO: Remove line before merging to mainline -->
+
+This is pending [AWS Cryptographers' guidance](https://t.corp.amazon.com/P148158466/communication).
+
+#### Option 2: Change the hashing algorithm to SHA384 in both
+
+SHA512 is the better choice as mentioned above.
+However, AWS cryptographer's generic recommendation for
+Hash Functions is SHA-384.
+We believe they make this recommendation based on criteria
+other than collision resistance.
+
+<!-- TODO: Once CryptoBR has confirmed SHA-512 >> SHA-384 for
+this purpose, revise this section to state "We know" instead of "We believe" -->
+
+More information on length extension attacks
+in SHA384 and SHA512 can be found
+[here in the appendix](#a3-information-about-sha384-and-length-extension-attacks).
+
+#### Option 3: Using different hashing algorithms
+
+This would mean different cache identifiers will be
+unnecesserily incoherent. Also this deviates from
+AWS Cryptographers' guidance.
+
+### Issue 4: Should we add a prefix to all cache identifiers in Hierarchy Keyrings and CachingCMMs?
+
+Here, by prefix we mean, adding the UTF8 encoding
+of the string 'hierarchy_keyring' as a
+prefix to a hierarchy keyring cache identifier,
+and adding encoding of 'caching_cmm' as a prefix to a
+CachingCMM cache identifier.
+
+#### Option 1 (Recommended): Yes
```

Comment:
I feel this is trivial. If you are not trusting your hash function then there are bigger issues, which is beyond adding a prefix.

Reply:
Updated the cache identifiers. This shouldn't be an issue now.